python/delta_sharing/reader.py

```diff
@@ -234,7 +203,146 @@ def to_pandas(self) -> pd.DataFrame:
 
         return merged[[col_map[field["name"].lower()] for field in schema_json["fields"]]]
 
+    def __write_temp_delta_log_snapshot(self, temp_dir: str, lines: Sequence[str]) -> str:
+        delta_log_dir_name = temp_dir
+        table_path = "file:///" + delta_log_dir_name
+
+        # Create a new directory named '_delta_log' within the temporary directory
+        log_dir = os.path.join(delta_log_dir_name, '_delta_log')
+        os.makedirs(log_dir)
+
+        # Create a new .json file within the '_delta_log' directory
+        json_file_name = "0".zfill(20) + ".json"
+        json_file_path = os.path.join(log_dir, json_file_name)
+        json_file = open(json_file_path, 'w+')
+
+        # Write the protocol action to the log file
+        protocol_json = loads(lines.pop(0))
+        deltaProtocol = {"protocol": protocol_json["protocol"]["deltaProtocol"]}
+        dump(deltaProtocol, json_file)
+        json_file.write("\n")
+
+        # Write the metadata action to the log file
+        metadata_json = loads(lines.pop(0))
+        deltaMetadata = {"metaData": metadata_json["metaData"]["deltaMetadata"]}
+        dump(deltaMetadata, json_file)
+        json_file.write("\n")
+
+        # Write the add file actions to the log file
+        for line in lines:
+            line_json = loads(line)
+            dump(line_json["file"]["deltaSingleAction"], json_file)
+            json_file.write("\n")
+
+        # Close the file
+        json_file.close()
+        return table_path
+
+    def __write_temp_delta_log_cdf(self, temp_dir: str, lines: Sequence[str], start_version: Optional[int]) -> tuple[str, int, int]:
+        # TODO: use with when opening files
+        delta_log_dir_name = temp_dir
+        table_path = "file:///" + delta_log_dir_name
+
+        # Create a new directory named '_delta_log' within the temporary directory
+        log_dir = os.path.join(delta_log_dir_name, '_delta_log')
+        os.makedirs(log_dir)
+
+        # first two lines are protocol and metadata, respectively
+        protocol_json = loads(lines.pop(0))
+        deltaProtocol = {"protocol": protocol_json["protocol"]["deltaProtocol"]}
+        metadata_json = loads(lines.pop(0))
+        deltaMetadata = {"metaData": metadata_json["metaData"]["deltaMetadata"]}
+
+        min_version = start_version if start_version is not None else (10**20 - 1)
+        max_version = 0
+        version_to_actions = defaultdict(list)
+        version_to_timestamp = {}
+        
+        # Construct map from version to actions that took place in that version
+        for line in lines:
+            line_json = loads(line)
+            file = line_json["file"]
```

Comment:
there could be metadata in lines, that needs to be handled.
https://github.com/delta-io/delta-sharing/blob/main/PROTOCOL.md#metadata-in-delta-format